### PR TITLE
gcov: fix clang build

### LINKED ISSUE #10612: Can't build with clang and gcov enabled
**Bug description**

Can't build tarantool with clang and gcov enabled.

* OS: NixOS
* OS Version: 24.11 (Vicuna)
* Architecture: amd64

```bash
git describe
3.3.0-entrypoint-106-g1080995fa4
```

**Steps to reproduce**

On a system with clang compiler run:
```bash
mkdir build && cd build && cmake -DENABLE_GCOV=ON ../
```

**Actual behavior**

```log
CMake Error at cmake/profile.cmake:17 (message):
  ENABLE_GCOV option requested but gcov library is not found
Call Stack (most recent call first):
  CMakeLists.txt:79 (include)
```

<details>

<summary>
Full log
</summary>

```log
CMake Deprecation Warning at CMakeLists.txt:1 (cmake_minimum_required):
  Compatibility with CMake < 3.5 will be removed from a future version of
  CMake.

  Update the VERSION argument <min> value or use a ...<max> suffix to tell
  CMake that the project does not need compatibility with older versions.


-- The C compiler identification is Clang 18.1.8
-- The CXX compiler identification is Clang 18.1.8
-- The ASM compiler identification is Clang with GNU-like command-line
-- Found assembler: /nix/store/4rz4z2bkb68vwbdxcwq0jxh2fyhhiqkh-clang-wrapper-18.1.8/bin/clang
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /nix/store/4rz4z2bkb68vwbdxcwq0jxh2fyhhiqkh-clang-wrapper-18.1.8/bin/clang - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /nix/store/4rz4z2bkb68vwbdxcwq0jxh2fyhhiqkh-clang-wrapper-18.1.8/bin/clang++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Building for Linux
-- Performing Test HAVE_STD_C11
-- Performing Test HAVE_STD_C11 - Success
-- Performing Test HAVE_STD_GNU99
-- Performing Test HAVE_STD_GNU99 - Success
-- Performing Test HAVE_STD_CXX11
-- Performing Test HAVE_STD_CXX11 - Success
-- Performing Test HAVE_STD_GNUXX0X
-- Performing Test HAVE_STD_GNUXX0X - Success
-- Performing Test CC_HAS_WNO_PARENTHESES
-- Performing Test CC_HAS_WNO_PARENTHESES - Success
-- Performing Test CC_HAS_WNO_PARENTHESES_EQUALITY
-- Performing Test CC_HAS_WNO_PARENTHESES_EQUALITY - Success
-- Performing Test CC_HAS_WNO_TAUTOLOGICAL_COMPARE
-- Performing Test CC_HAS_WNO_TAUTOLOGICAL_COMPARE - Success
-- Performing Test CC_HAS_WNO_MISLEADING_INDENTATION
-- Performing Test CC_HAS_WNO_MISLEADING_INDENTATION - Success
-- Performing Test CC_HAS_WNO_VARARGS
-- Performing Test CC_HAS_WNO_VARARGS - Success
-- Performing Test CC_HAS_WNO_IMPLICIT_FALLTHROUGH
-- Performing Test CC_HAS_WNO_IMPLICIT_FALLTHROUGH - Success
-- Performing Test CC_HAS_WNO_CAST_FUNCTION_TYPE
-- Performing Test CC_HAS_WNO_CAST_FUNCTION_TYPE - Success
-- Performing Test CC_HAS_GGDB
-- Performing Test CC_HAS_GGDB - Success
-- Performing Test HAVE_FORCE_ALIGN_ARG_POINTER_ATTR
-- Performing Test HAVE_FORCE_ALIGN_ARG_POINTER_ATTR - Success
-- Performing Test HAVE_CFI_ASM
-- Performing Test HAVE_CFI_ASM - Success
-- Looking for stdatomic.h
-- Looking for stdatomic.h - found
-- Enabling LTO: FALSE
-- Performing Test CC_HAS_SSE2_INTRINSICS
-- Performing Test CC_HAS_SSE2_INTRINSICS - Success
-- Performing Test CC_HAS_AVX_INTRINSICS
-- Performing Test CC_HAS_AVX_INTRINSICS - Success
-- SSE2 is enabled - target CPU must supppot it
-- Performing Test CC_HAS_ATOMICS
-- Performing Test CC_HAS_ATOMICS - Success
-- Looking for __gcov_flush in
-- Looking for __gcov_flush in  - not found
CMake Error at cmake/profile.cmake:17 (message):
  ENABLE_GCOV option requested but gcov library is not found
Call Stack (most recent call first):
  CMakeLists.txt:79 (include)
```

-- Configuring incomplete, errors occurred!

</details>

**Expected behavior**

cmake finishes successfully, tarantool can be built with gcov support afterwards.

### PR DESCRIPTION
**Pull request by grafin**
Fix the macros and cmake recipe for gcov to make it possible to build tarantool with clang compiler and gcov enabled.

Closes #10612

### CODE REVIEW

**Review by Totktonada — APPROVED** [review]
I'm not familiar with this machinery, to be honest. If it works, it is likely OK :)

File: cmake/profile.cmake [hunk 1]
```diff
@@ -1,5 +1,5 @@
 set(CMAKE_REQUIRED_FLAGS "-fprofile-arcs -ftest-coverage")
-if (CMAKE_C_COMPILER_ID STREQUAL "GNU" AND NOT CMAKE_C_COMPILER_VERSION VERSION_LESS 11)
+if (CMAKE_C_COMPILER_VERSION VERSION_GREATER_EQUAL 11)
```
Comment (ligurio) [inline cmake/profile.cmake:2]: For GCC the reason is described in commit message bd813168467e4ac614439f56126cbce4134d493c, for Clang the reason why the version should be greater than 11 is not clear.
I propose to add a comment for both GCC and Clang and probably change condition to

`(compiler == GCC && version >= 11) or (compiler == Clang)`

File: src/trivia/util.h [hunk 2]
```diff
@@ -542,7 +542,8 @@ alloc_failure(const char *filename, int line, size_t size)
 
 void close_all_xcpt(int fdc, ...);
 
-#if defined(__GNUC__) && __GNUC__ >= 11
+#if (defined(__GNUC__) && __GNUC__ >= 11) || \
+	(defined(__clang__) && __clang_major__ >= 11)
```
Comment (ligurio) [inline src/trivia/util.h:546]: The same: what is specific in Clang 11?

**Review by ligurio — COMMENTED** [review]
Borislav, thanks for the patch!
I left a couple of comments, please take a look.

File: cmake/profile.cmake (hunk 1 above)
Comment (grafin) [reply cmake/profile.cmake:2]: Ok, I've did some research) and the same change as in GCC 11 was introduced in llvm 12. https://github.com/llvm/llvm-project/commit/5809a32e7c2d79a9a463eb9c15cde994b42e3002
I've updated the PR to check for that and tested it with clang 12 and 18. I tried getting a clang 11 enviroment for my nixos, but it has been removed from repositories: https://github.com/NixOS/nixpkgs/pull/283591. If you insist n testing - I could get some old container with clang 11 and test it there, but rather wouldn't)

Comment (grafin) [reply cmake/profile.cmake:2]: Ok, I've did some research) and the same change as in GCC 11 was introduced in llvm 12. https://github.com/llvm/llvm-project/commit/5809a32e7c2d79a9a463eb9c15cde994b42e3002
I've updated the PR to check for that and tested it with clang 12 and 18. I tried getting a clang 11 enviroment for my nixos, but it has been removed from repositories: https://github.com/NixOS/nixpkgs/pull/283591. If you insist n testing - I could get some old container with clang 11 and test it there, but rather wouldn't)

Comment (ligurio) [reply cmake/profile.cmake]: Thanks for the fix!

> Ok, I've did some research) and the same change as in GCC 11 was introduced in llvm 12. https://github.com/llvm/llvm-project/commit/5809a32e7c2d79a9a463eb9c15cde994b42e3002

I would add a link at least to the commit message.

**Review by ligurio — APPROVED** [review]
LGTM

File: cmake/profile.cmake (hunk 1 above)
Comment (grafin) [reply cmake/profile.cmake]: Done!

**Pull request merged**